Multiple file metainfo mode support in cmd tools

milosgajdos83 opened the pull request:
```torrent-create``` and ```torrent-verify``` did not work correctly when a single file is present in ```metainfo``` torrent file. ```torrent-create``` now correctly fails if the path passed as an command line argument does not exist - otherwise it throws runtime panic as the non-existent file case is not handled by it. ```torrent-verify``` now has a summary flag. Finally, in ```span.go``` I've cheekily sneaked in a change which defines the ```span``` type more like idiomatic Go (I have a weird OCD :-)).

anacrolix commented on cmd/torrent-create/main.go:
```diff
@@ -26,6 +26,9 @@ func init() {
+		if _, err := os.Stat(filename); os.IsNotExist(err) {
+			log.Fatal(err)
+		}
 		if err := filepath.Walk(filename, func(path string, info os.FileInfo, err error) error {
```
I think you've introduced a race via LBYL. You should check the error inside the Walk? This is the reason for the panic.

milosgajdos83 replied:
I tried to avoid the ```Walk``` completely if the ```root``` path does not exist. I'll move this to the ```Walk``` now.

anacrolix commented on cmd/torrent-verify/main.go:
```diff
@@ -16,16 +16,40 @@ import (
+	fmt.Println(" TORRENT-VERIFY ")
+	fmt.Println("----------------")
+	fmt.Printf("Number of correct pieces: %d\n", len(sMap[true]))
+	fmt.Printf("Number of wrong pieces: %d\n", len(sMap[false]))
```
I think you could just blow away what's currently logged, and show a progress line with something like: X% complete (X/Y pieces correct). When you finish, print a newline?

milosgajdos83 replied:
The reason why I added the summary flag is, if you have a file with thousands of pieces they're logged in the ```stdout``` and unless you parse them with some UNIX cli tool magic it can get pretty verbose. 
The new ```summary``` flag is optional and just gives a quick overview only if you wish to see one. I'm happy to remove it completely.

anacrolix replied:
I'm suggesting you ditch the old way, and just go with this. But print a nice progress bar as it goes. Are you familiar with the carriage return character? I did a crude form of it in cmd/torrent: https://github.com/anacrolix/torrent/blob/master/cmd/torrent/main.go#L150

milosgajdos83 replied:
I kinda like the original idea of having a visibility of which piece is incorrect.  I'm happy to implement the progress bit, but like I said we will lose the per-piece visibility, which does not necessarily have to be a bad thing ?

anacrolix commented on cmd/torrent-verify/main.go:
```diff
@@ -16,16 +16,40 @@ import (
+		log.Printf("file mmap has wrong size: %#v", filename)
+	}
+	osFile.Close()
+	mMapSpan.Append(goMMap)
```
Simpler to just return the gommap.MMap, and append it in the caller.

milosgajdos83 replied:
Yep, great suggestion. That way the function name would also be more correct semantically :+1:

anacrolix commented on dht/dht.go:
```diff
@@ -980,11 +980,13 @@ func bootstrapAddrs(nodeAddrs []string) (addrs []*net.UDPAddr, err error) {
-		udpAddr, err := net.ResolveUDPAddr("udp4", addrStr)
-		if err != nil {
-			continue
+		if addrStr != "" {
```
I don't have a name for this, but it's like passing the buck on a sentinel value. You've created a new special case which is addrStr=="". Maybe it could mean something. Maybe not. It's not our problem unless we care about the errors we're getting, which we don't. What do you think of that?

milosgajdos83 replied:
I agree with this. It is an ugly hack :-) The reason why I have added the check is, when you are starting your own bittorrent DHT and want to avoid the external ones (ie. utorrent.com and bittorrent.com), currently the ```bootstrapNodes``` must have at least one item.

I need to think about a way to start the "private/internal" bittorrent DHT in some cleaner way than this terrible hack :+1: 
Do you have any suggestions here ?

anacrolix commented on mmap_span/span.go:
```diff
@@ -4,9 +4,7 @@ type sizer interface {
-type (
-	span []sizer
-)
+type span []sizer
```
Nice catch.

anacrolix commented on cmd/torrent-verify/main.go:
```diff
@@ -16,16 +16,40 @@ import (
-	filePath = flag.String("torrent", "/path/to/the.torrent", "path of the torrent file")
-	dirPath  = flag.String("path", "/torrent/data", "path of the torrent data")
+	torrentPath = flag.String("torrent", "/path/to/the.torrent", "path of the torrent file")
+	dataPath    = flag.String("path", "/torrent/data", "path of the torrent data")
```
Currently torrent-verify works by verifying the data in the current working directory. The OS already has a provision for setting the working directory, I think adding another flag just complicates the the cmd's interface. A doc note saying "data is read from the working directory" could be better.

anacrolix commented:
I appreciate the improvements, keep them coming, thanks.

milosgajdos83 commented:
I've addressed some of your comments. I think we can leave the summary bit out. It does not really improve much to be honest. I'd like to keep the path to data flag called ```dataPath``` instead of ```dirPath```. I need to think about that DHT stuff - I've removed my ugly hack :-) 
Thanks for the proper code reviews!

anacrolix commented:
Great work.
